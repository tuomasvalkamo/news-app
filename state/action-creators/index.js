export const addBookmark = (bookmark) => {
  return (dispatch) => {
    dispatch({
      type: "ADD_BOOKMARK",
      payload: {
        objectID: bookmark.objectID,
        title: bookmark.title,
        date: bookmark.date,
        url: bookmark.url,
        formattedUrl: bookmark.formattedUrl
      }
    })
  }
}

export const removeBookmark = (objectID) => {
  return (dispatch) => {
    dispatch({
      type: "REMOVE_BOOKMARK",
      payload: {
        objectID: objectID
      }
    })
  }
}

export const setBookmarksFromDB = (bookmarks) => {
  return (dispatch) => {
    dispatch({
      type: "SET_BOOKMARKS_FROM_DB",
      payload: {
        bookmarks: bookmarks
      }
    })
  }
}