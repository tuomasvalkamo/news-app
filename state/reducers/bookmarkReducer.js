const reducer = (state = [], action) => {
  switch (action.type) {
    case "SET_BOOKMARKS_FROM_DB":
      return action.payload.bookmarks
    case "ADD_BOOKMARK":
      return [
        ...state,
        {
          objectID: action.payload.objectID,
          title: action.payload.title,
          date: action.payload.date,
          url: action.payload.url,
          formattedUrl: action.payload.formattedUrl,
        }
      ]
    case "REMOVE_BOOKMARK":
      return state.filter(bookmark => bookmark.objectID !== action.payload.objectID)
    default:
      return state
  }
}

export default reducer