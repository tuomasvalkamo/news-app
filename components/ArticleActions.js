import React, { useState, useEffect } from 'react'
import { View, Pressable, StyleSheet } from 'react-native'
import { Ionicons } from '@expo/vector-icons'
import { saveBookmark, deleteBookmark } from './AsyncStorage';

import { useSelector, useDispatch } from 'react-redux'
import { bindActionCreators } from 'redux';
import { actionCreators } from '../state/index';

function ArticleActions(props) {
  const [bookmarked, setBookmarked] = useState(false)
  const [isDeleted, setIsDeleted] = useState(false)

  // Redux state management
  const bookmarks = useSelector(state => state.bookmarks)
  const dispatch = useDispatch()
  const { addBookmark, removeBookmark } = bindActionCreators(actionCreators, dispatch)

  useEffect(() => {
    isDeleted ? '' : checkIfBookmarked()
  }, [bookmarks])

  const checkIfBookmarked = () => {
    if (bookmarks) {
      for (let i = 0; i < bookmarks.length; i++) {
        if (bookmarks[i].url === props.url) {
          setBookmarked(true)
        }
      }
    }
  }

  const handleBookmarkPress = () => {
    if (!bookmarked) {
      setBookmarked(true)

      // Save bookmark to AsyncStorage
      saveBookmark(props.objectID, props.title, props.date, props.url, props.formattedUrl)

      // Add bookmark to Redux state
      const newBookmarkObject = {
        objectID: props.objectID,
        title: props.title,
        date: props.date,
        url: props.url,
        formattedUrl: props.formattedUrl
      }

      addBookmark(newBookmarkObject)
    } else {
      setIsDeleted(true)
      setBookmarked(false)
      // Delete bookmark from AsyncStorage
      deleteBookmark(props.objectID)
      // Remove bookmark from Redux state
      removeBookmark(props.objectID)
    }
  }

  return (
    <View style={styles.infoActions}>
      <Pressable
        onPress={() => console.log('asd')}
      >
        <Ionicons name='share-social-outline' size={24} style={styles.share} />
      </Pressable>
      <Pressable
        onPress={() => handleBookmarkPress()}
      >
        <Ionicons name={bookmarked === true ? 'bookmark' : 'bookmark-outline'} size={24} />
      </Pressable>
    </View>
  )
}

const styles = StyleSheet.create({
  infoActions: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'flex-end'
  },
  share: {
    marginRight: 5
  }
})

export default ArticleActions
