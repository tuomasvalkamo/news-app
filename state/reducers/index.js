import { combineReducers } from "redux";
import bookmarkReducer from "./bookmarkReducer";

const reducers = combineReducers({
  bookmarks: bookmarkReducer
})

export default reducers