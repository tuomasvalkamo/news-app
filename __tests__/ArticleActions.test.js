import React from 'react'
import { create, act } from 'react-test-renderer'
import ArticleActions from '../components/ArticleActions'
import { createStore, applyMiddleware, bindActionCreators } from 'redux'
import { Provider, useSelector, useDispatch } from 'react-redux'
import reducers from '../state/reducers/index'
import thunk from 'redux-thunk'
import { actionCreators } from '../state/index';

const store = createStore(
  reducers,
  {},
  applyMiddleware(thunk)
)

const tree = create(
  <Provider store={store}>
    <ArticleActions />
  </Provider>
)

test('renders correctly', () => {
  expect(tree).toMatchSnapshot();
});
/*
test('bookmark is added to redux', async () => {
  // press button
  const bookmarkPressable =
    tree.root.findByProps({ testID: 'sharePressable' }).props;
  act(() => bookmarkPressable.onPress())
  // expect redux storage to have bookmark
  console.log(store.getState())
  expect(store.getState().bookmarks.length).toEqual(1)
})*/