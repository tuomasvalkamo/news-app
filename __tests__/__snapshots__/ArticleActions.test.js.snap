// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-end",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "flex-end",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="sharePressable"
  >
    <Text />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="bookmarkPressable"
  >
    <Text />
  </View>
</View>
`;
